server/src/main/java/org/opensearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -1194,51 +1045,12 @@ private void cleanFiles(
             ActionListener.delegateResponse(listener, (l, e) -> ActionListener.completeWith(l, () -> {
                 StoreFileMetadata[] mds = StreamSupport.stream(sourceMetadata.spliterator(), false).toArray(StoreFileMetadata[]::new);
                 ArrayUtil.timSort(mds, Comparator.comparingLong(StoreFileMetadata::length)); // check small files first
-                handleErrorOnSendFiles(store, e, mds);
+                transferHandler.handleErrorOnSendFiles(store, e, mds);
                 throw e;
             }))
         );
     }
 
-    private void handleErrorOnSendFiles(Store store, Exception e, StoreFileMetadata[] mds) throws Exception {
-        final IOException corruptIndexException = ExceptionsHelper.unwrapCorruption(e);
-        assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[handle error on send/clean files]");
-        if (corruptIndexException != null) {
-            Exception localException = null;
-            for (StoreFileMetadata md : mds) {
-                cancellableThreads.checkForCancel();
-                logger.debug("checking integrity for file {} after remove corruption exception", md);
-                if (store.checkIntegrityNoException(md) == false) { // we are corrupted on the primary -- fail!
-                    logger.warn("{} Corrupted file detected {} checksum mismatch", shardId, md);
-                    if (localException == null) {
-                        localException = corruptIndexException;
-                    }
-                    failEngine(corruptIndexException);
```

Comment:
It seems like `failEngine()` is called only in this flow, why we keep the method there + using the method reference in `transferHandler` instead of just moving the method to `SegmentFileTransferHandler`?

Reply:
I did this so that I wouldn't have to make any changes to RecoverySourceHandlerTests to prove I didn't break anything there.  I agree though this should just call shard.fail directly, will clean this up and move the necessary tests over.

Reply:
Updating `RecoverySourceHandlerTests` wasn't as bad as I anticipated, also added a separate test class for `SegmentFileTransferHandler`.